File: app/src/main/java/org/ametro/routes/MapRouteProvider.java
package org.ametro.routes;

import org.ametro.model.entities.MapTransportLine;
import org.ametro.model.entities.MapTransportScheme;
import org.ametro.model.entities.MapTransportSegment;
import org.ametro.model.entities.MapTransportTransfer;
import org.ametro.routes.algorithms.DijkstraHeap;
import org.ametro.routes.entities.MapRoute;
import org.ametro.routes.entities.MapRoutePart;
import org.ametro.routes.entities.MapRouteQueryParameters;

import java.util.ArrayList;
import java.util.Collections;

public class MapRouteProvider {

    public static MapRoute[] findRoutes(MapRouteQueryParameters parameters) {

        DijkstraHeap.TransportGraph graph = new DijkstraHeap.TransportGraph(parameters.getStationCount());

        CreateGraphEdges(graph, parameters);

        DijkstraHeap.Result result = DijkstraHeap.dijkstra(graph, parameters.getBeginStationUid());

        if(result.getPredecessors()[parameters.getEndStationUid()] == DijkstraHeap.NO_WAY){
            return new MapRoute[0];
        }

        return new MapRoute[]{ ConvertToMapRoute(result, parameters.getEndStationUid()) };
    }

    private static MapRoute ConvertToMapRoute(DijkstraHeap.Result result, int endStationUid) {
        final long[] distances = result.getDistances();
        final int[] predecessors = result.getPredecessors();

        ArrayList<MapRoutePart> parts = new ArrayList<>();
        int _to = endStationUid;
        int _from = predecessors[_to];

        while (_from != DijkstraHeap.NO_WAY) {
            parts.add(new MapRoutePart(_from, _to, distances[_to] - distances[_from]));
            _to = _from;
            _from = predecessors[_to];
        }

        Collections.reverse(parts);

        return new MapRoute(parts.toArray(new MapRoutePart[parts.size()]));
    }

    private static void CreateGraphEdges(DijkstraHeap.TransportGraph graph, MapRouteQueryParameters parameters) {
        MapTransportLine[] stationLines = new MapTransportLine[parameters.getStationCount()];
        for (MapTransportScheme scheme : parameters.getEnabledTransportsSchemes()) {
            for (MapTransportLine line : scheme.getLines()) {
                for (MapTransportSegment segment : line.getSegments()) {
                    if (segment.getDelay() == 0) {
                        continue;
                    }
                    graph.addEdge(segment.getFrom(), segment.getTo(), segment.getDelay());
                    stationLines[segment.getFrom()] = line;
                    stationLines[segment.getTo()] = line;
                }
            }
            final Integer delayIndex = parameters.getDelayIndex();
            for (MapTransportTransfer transfer : scheme.getTransfers()) {

                int delay = transfer.getDelay() ;
                if(delayIndex!=null && stationLines[transfer.getTo()]!=null){

                    final int[] lineDelays = stationLines[transfer.getTo()].getDelays();
                    if(lineDelays.length!=0) {
                        delay += stationLines[transfer.getTo()].getDelays()[delayIndex];
                    }
                }

                graph.addEdge(transfer.getFrom(), transfer.getTo(), delay);
                graph.addEdge(transfer.getTo(), transfer.getFrom(), delay);
            }
        }
    }
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/routes/MapRouteProvider.kt
package io.github.romangolovanov.apps.ametro.routes

import io.github.romangolovanov.apps.ametro.routes.algorithms.DijkstraHeap
import io.github.romangolovanov.apps.ametro.routes.entities.MapRoute
import io.github.romangolovanov.apps.ametro.routes.entities.MapRoutePart
import io.github.romangolovanov.apps.ametro.routes.entities.MapRouteQueryParameters
import java.util.Collections

object MapRouteProvider {

    @JvmStatic
    fun findRoutes(parameters: MapRouteQueryParameters): Array<MapRoute> {
        val graph = DijkstraHeap.TransportGraph(parameters.stationCount)
        createGraphEdges(graph, parameters)
        val result = DijkstraHeap.dijkstra(graph, parameters.beginStationUid)

        if (result.predecessors[parameters.endStationUid] == DijkstraHeap.NO_WAY) {
            return emptyArray()
        }
        return arrayOf(convertToMapRoute(result, parameters.endStationUid))
    }

    private fun convertToMapRoute(result: DijkstraHeap.Result, endStationUid: Int): MapRoute {
        val distances = result.distances
        val predecessors = result.predecessors

        val parts = mutableListOf<MapRoutePart>()
        var to = endStationUid
        var from = predecessors[to]

        while (from != DijkstraHeap.NO_WAY) {
            parts.add(MapRoutePart(from, to, distances[to] - distances[from]))
            to = from
            from = predecessors[to]
        }

        Collections.reverse(parts)
        return MapRoute(parts.toTypedArray())
    }

    private fun createGraphEdges(graph: DijkstraHeap.TransportGraph, parameters: MapRouteQueryParameters) {
        val stationLines = arrayOfNulls<io.github.romangolovanov.apps.ametro.model.entities.MapTransportLine>(parameters.stationCount)

        for (scheme in parameters.getEnabledTransportsSchemes()) {
            for (line in scheme.lines) {
                for (segment in line.segments) {
                    if (segment.delay == 0) continue
                    graph.addEdge(segment.from, segment.to, segment.delay)
                    stationLines[segment.from] = line
                    stationLines[segment.to] = line
                }
            }
            val delayIndex = parameters.delayIndex
            for (transfer in scheme.transfers) {
                var delay = transfer.delay
                if (delayIndex != null && stationLines[transfer.to] != null) {
                    val lineDelays = stationLines[transfer.to]!!.delays
                    if (lineDelays.isNotEmpty()) {
                        delay += lineDelays[delayIndex]
                    }
                }
                graph.addEdge(transfer.from, transfer.to, delay)
                graph.addEdge(transfer.to, transfer.from, delay)
            }
        }
    }
}
